crates/data_components/src/s3_single_file_cached.rs

```diff
@@ -0,0 +1,231 @@
+        let table_paths = listing_table.table_paths();
+
+        // Only wrap single-file tables
+        if table_paths.len() != 1 || table_paths[0].is_collection() {
+            return None;
+        }
+
+        let file_path = ObjectStorePath::from(table_paths[0].prefix().as_ref());
+
+        Some(Self {
+            inner: listing_table,
+            object_store,
+            file_path,
+            cached_metadata: Arc::new(RwLock::new(None)),
+            dataset_name,
+        })
+    }
+
+    /// Fetches the current metadata (`ETag`, Version ID, size, `last_modified`) for the file from S3.
+    async fn fetch_current_metadata(&self) -> DataFusionResult<Option<object_store::ObjectMeta>> {
+        match self.object_store.head(&self.file_path).await {
+            Ok(meta) => Ok(Some(meta)),
+            Err(e) => {
+                tracing::debug!(
+                    "Failed to fetch S3 file metadata for {}: {}",
+                    self.dataset_name,
+                    e
+                );
+                Ok(None)
+            }
+        }
+    }
+
+    /// Checks if the file's metadata has changed since the last scan.
+    /// Returns `true` if the file is unchanged and can be skipped.
+    async fn is_file_unchanged(&self) -> DataFusionResult<bool> {
+        let Some(current_metadata) = self.fetch_current_metadata().await? else {
+            return Ok(false); // Can't determine, assume changed
+        };
+
+        let cached = self.cached_metadata.read().await;
+
+        if let Some(cached_meta) = cached.as_ref() {
+            // Check if refresh should be skipped:
+            // Skip if size, last_modified, AND version info all match
+            if current_metadata.size == cached_meta.size
+                && current_metadata.last_modified == cached_meta.last_modified
+                && is_same_file_version(cached_meta, &current_metadata)
+            {
+                tracing::debug!(
+                    "Skipping refresh for {} (file unchanged at {})",
+                    self.dataset_name,
+                    current_metadata.location
+                );
+                return Ok(true);
+            }
+        }
+
+        // Update cache with new metadata
+        drop(cached);
+        *self.cached_metadata.write().await = Some(current_metadata);
+
+        Ok(false)
+    }
+
+    /// Public method to check if a refresh should be skipped for this file.
+    /// Returns `true` if the file's metadata (`ETag`, Version, size, timestamp) hasn't changed.
+    /// This should be called during refresh operations before fetching data.
+    pub async fn should_skip_refresh(&self) -> DataFusionResult<bool> {
+        self.is_file_unchanged().await
+    }
+}
+
+#[async_trait]
+impl RefreshSkipTableProvider for S3SingleFileCached {
+    async fn should_skip_refresh(&self) -> DataFusionResult<bool> {
+        Ok(self.is_file_unchanged().await?)
+    }
+}
+
+#[async_trait]
+impl TableProvider for S3SingleFileCached {
```

Comment:
Implement all of the optional methods to delegate to the inner. This has been a huge source of bugs for us in the past